File: 2021/day_12.py
from collections import defaultdict

def load_input(filename="input.txt"):
  with open(filename) as f_obj:
    data = [x.strip() for x in f_obj.readlines()]
  return data

def parse_data(data):
  caves = defaultdict(lambda: [])
  for line in data:
    a, b = line.split("-")

    caves[a].append(b)
    caves[b].append(a)

  return caves

def find_routes(caves, start, end, small_seen=[]):
  if start == end:
    return ["end",]
  else:
    routes = []
    for opt in caves[start]:
      valid = True
      if opt == 'start' or (opt.lower() == opt and opt in small_seen):
        valid = False
      
      if valid:
        new_ss = [x for x in small_seen]
        if opt.lower() == opt:
          new_ss.append(opt)
        new_routes = find_routes(caves, opt, end, new_ss)
        for x in new_routes:
          if type(x) == str:
            routes.append([start, x])
          else:
            y = [start]
            y.extend(x)
            routes.append(y)
    return routes
  
def find_double_routes(caves, start, end, small_seen={}):
  if start == end:
    return ["end",]
  else:
    routes = []
    for opt in caves[start]:
      valid = True
      if opt == 'start' or (opt.lower() == opt and opt in small_seen and any([val > 1 for val in small_seen.values()])):
        valid = False
      
      if valid:
        new_ss = {k:v for k,v in small_seen.items()}
        if opt.lower() == opt:
          if opt in new_ss:
            new_ss[opt] += 1
          else:
            new_ss[opt] = 1
        new_routes = find_double_routes(caves, opt, end, new_ss)
        for x in new_routes:
          if type(x) == str:
            routes.append([start, x])
          else:
            y = [start]
            y.extend(x)
            routes.append(y)
    return routes

def first_star(data):
  caves = parse_data(data)
  routes = find_routes(caves, "start", "end", small_seen=[])
  return len(routes)

def second_star(data):
  caves = parse_data(data)
  routes = find_double_routes(caves, "start", "end", small_seen=dict())
  return len(routes)

if __name__ == "__main__":
  data = load_input()
  print("First star:", first_star(data))
  print("Second star:", second_star(data))

File: 2021/day_12.rs
use std::fs;
use std::collections::HashMap;

fn load_data(filepath:&str) -> String {
  let file = fs::read_to_string(filepath);
  if file.is_ok() {
    return file.unwrap();
  }
  panic!("File not found or unreadable");
  }

fn parse_caves(data:&String) -> HashMap<String, Vec<String>> {
  let mut caves:HashMap<String, Vec<String>> = HashMap::new();

  for route in data.split('\n') {
    let parts:Vec<String> = route.split('-').map(|s| String::from(s)).collect();
    if parts[1] != String::from("start") {
          let c = caves.entry(parts[0].clone()).or_insert(Vec::new());
      c.push(parts[1].clone());
    }
    if parts[0] != String::from("start") {
      let d = caves.entry(parts[1].clone()).or_insert(Vec::new());
      d.push(parts[0].clone())
    }
  }

  caves.remove("end");

  return caves;
}

fn find_routes(caves:&HashMap<String, Vec<String>>, start:String, end:String, small_seen:Vec<String>) -> Vec<Vec<String>> {
// Find all valid routes from the start to the end;
// Probably some recursive/memoised nonsense.

if start == end {
  let mut end_vec = Vec::new();
  end_vec.push(vec!["end".to_string()]);
  return end_vec;
} else {
  let options:Vec<String> = caves.get(&start).unwrap().iter().map(|s| String::from(s)).collect();
  let mut routes:Vec<Vec<String>> = Vec::new();

  for opt in options {
    let mut valid = true;

    if opt == "start" || small_seen.contains(&opt) {
      valid = false;
    }

    if valid {
      let mut new_ss = small_seen.clone();
      if opt.to_lowercase() == opt {
        new_ss.push(opt.clone());
      }
      let new_routes = find_routes(&caves, opt, end.clone(), new_ss);

      for mut x in new_routes {
        x.insert(0, start.clone());
        routes.push(x);
      }
    }
  }

  return routes;
}
}

fn first_star(data:&String) -> i32 {

  let caves = parse_caves(data);

  let routes = find_routes(&caves, String::from("start"), String::from("end"), Vec::new());

  return routes.len() as i32;
}

fn find_double_routes(caves:&HashMap<String, Vec<String>>, start:String, end:String, small_seen:HashMap<String, i32>) -> Vec<Vec<String>> {
// Find all valid routes from the start to the end;
// Probably some recursive/memoised nonsense.

if start == end {
  let mut end_vec = Vec::new();
  end_vec.push(vec!["end".to_string()]);
  return end_vec;
} else {
  let options:Vec<String> = caves.get(&start).unwrap().iter().map(|s| String::from(s)).collect();
  let mut routes:Vec<Vec<String>> = Vec::new();

  for opt in options {
    let mut valid = true;

    if opt == "start" {
      valid = false;
    }

    if small_seen.contains_key(&opt) {
      for (k, v) in &small_seen {
        if v > &1 {
          valid = false;
        }
      }
    }

    if valid {
      let mut new_ss = small_seen.clone();
      if opt.to_lowercase() == opt {
        let o = new_ss.entry(opt.clone()).or_insert(0);
        *o += 1;
      }
      let new_routes = find_double_routes(&caves, opt, end.clone(), new_ss);

      for mut x in new_routes {
        x.insert(0, start.clone());
        routes.push(x);
      }
    }
  }

  return routes;
}
}

fn second_star(data:&String) -> i32 {
  let caves = parse_caves(data);

  let routes = find_double_routes(&caves, String::from("start"), String::from("end"), HashMap::new());

  return routes.len() as i32;
}

fn main() {
  let data = load_data("input.txt");
  println!("First star: {}", first_star(&data));
  println!("Second star: {}", second_star(&data));
}
